cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/GlobalRemoteAuthenticatorProvider.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.common.internal.remote;
+
+import io.cdap.cdap.security.spi.authenticator.RemoteAuthenticator;
+
+import javax.inject.Provider;
+
+/**
+ * Represents a singleton which provides a remote authenticator.
+ */
+public interface GlobalRemoteAuthenticatorProvider extends Provider<RemoteAuthenticator> {
+  /**
+   * Sets the authenticator to use. This applies for a given service until the
+   * {@link GlobalRemoteAuthenticatorProvider#resetAuthenticator()} method is called.
+   * @param authenticatorName The name of the authenticator to use.
+   */
+  void setAuthenticator(String authenticatorName);
```

Comment:
I afraid a mutable interface is not a very good design. What do we want to achieve here?

Reply:
Originally wanted to avoid refactoring too much since the original RemoteAuthenticator used static methods to maintain the current remote authenticator instance in a static variable, and changing the existing pattern was extremely brittle and prone to bugs.

However, I've gone ahead and refactored the entire RemoteAuthenticator interface to instead directly use a provider and inject based on a Guice binding at injector creation time instead. PTAL, thanks!